### ISSUE TITLE
podman network ls format doesn't work

### ISSUE THREAD

**Issue by tmds** [containers/podman#7865]
`podman network ls --format` doesn't work:

```
$ podman network ls  --format {{.ID}}
Error: template: listNetworks:1:14: executing "listNetworks" at <.ID>: can't evaluate field ID in type network.ListPrintReports
```

version:
```
podman.x86_64                                                         2:2.1.0-2.fc32                                                          @updates-testing
```

**Comment by dzerhusen** [comment]
Where do you got the `{{.ID}}` from? I must say, i am only using version 2.0.4. 

`podman network ls --format {{.Name}}`
works for example.

**Comment by tmds** [comment]
> podman network ls --format {{.Name}}

That works for me too. I got the `{{.ID}}` by copy-paste.

Do you know how I can see the full json?

**Comment by zhangguanzhang** [comment]
```
./bin/podman network ls --format json
```
there is no filed named `ID`

**Comment by mheon** [comment]
I believe this is a difference between us and Docker for now. We have plans
for a partial rewrite of the 'podman network's commands to address this but
it could take a while before we get to it.

**Comment by tmds** [comment]
I had tried, `--format {{json .}}`. I'm not sure if that is supposed to work.

@mheon should I close this issue? or do you want to keep it open?

**Comment by mheon** [comment]
Hm. I'll tag it as a feature request and it can stay open.

**Comment by rhatdan** [comment]
Hopefully someone from the community can take this on.

**Comment by eiffel-fl** [comment]
I can take a look to this issue.
My first research leads me to `type NetworkConfigList` in `libcni`.

I will finish to setup my development environment for `podman` and I will investigate more.

**Comment by rhatdan** [comment]
You got it.  Ping us if you need any help.

**Comment by eiffel-fl** [comment]
Maybe I underestimate the problem.
At first sight I just though the ID was not printed, so adding `ID string` to `type NetworkConfigList` should have solve the problem.
It solves it partly, indeed there is no more error.
But the ID is empty and I have the feeling that networks do not have ids in podman.
Even using the REST API through `curl` does not print the ID ([see](https://framabin.org/p/?1f5e399bc3e69914#qqNwbzQHLgdyK2pw2LWyCWhqhSZgd7GQQyTQ72kBABk=)).

I can try to add ID to networks but the modifications will be deeper and take more time.
If you have advices so my modifications don't break anything and be as restrained as possible I will take them.

**Comment by mheon** [comment]
You are correct that we do not presently have IDs for networks and cannot easily add them. It's on the list for the networking rewrite I'd like to do eventually, but that won't be done for a while.

Submitting anything you have right now seems like a good idea, given that.

**Comment by eiffel-fl** [comment]
I managed to add creation of ID with `stringid.GenerateNonCryptoID()` in `NewNcList`.
I also add the ID to `NetworkConfigList` and the ID is read from the file in `ConfListFromBytes`.
So the (long) ID are displayed when using `./bin/podman network ls --format '{{.ID}}'`.

Instead of opening a PR I think this is better if I share the patch here:
```
diff --git a/cmd/podman/networks/list.go b/cmd/podman/networks/list.go
index b6fb2bb80..f95e169f7 100644
--- a/cmd/podman/networks/list.go
+++ b/cmd/podman/networks/list.go
@@ -30,8 +30,8 @@ var (
 
 var (
 	networkListOptions entities.NetworkListOptions
-	headers            = "NAME\tVERSION\tPLUGINS\n"
-	defaultListRow     = "{{.Name}}\t{{.Version}}\t{{.Plugins}}\n"
+	headers            = "ID\tNAME\tVERSION\tPLUGINS\n"
+	defaultListRow     = "{{.ID}}\t{{.Name}}\t{{.Version}}\t{{.Plugins}}\n"
 )
 
 func networkListFlags(flags *pflag.FlagSet) {
diff --git a/pkg/network/netconflist.go b/pkg/network/netconflist.go
index 8187fdb39..a9835f113 100644
--- a/pkg/network/netconflist.go
+++ b/pkg/network/netconflist.go
@@ -4,6 +4,7 @@ import (
 	"net"
 	"os"
 	"path/filepath"
+	"github.com/containers/storage/pkg/stringid"
 )
 
 const (
@@ -20,6 +21,7 @@ func NewNcList(name, version string) NcList {
 	n := NcList{}
 	n["cniVersion"] = version
 	n["name"] = name
+	n["id"] = stringid.GenerateNonCryptoID()
 	return n
 }
 
diff --git a/vendor/github.com/containernetworking/cni/libcni/api.go b/vendor/github.com/containernetworking/cni/libcni/api.go
index 7e52bd838..a3695ef94 100644
--- a/vendor/github.com/containernetworking/cni/libcni/api.go
+++ b/vendor/github.com/containernetworking/cni/libcni/api.go
@@ -63,6 +63,7 @@ type NetworkConfig struct {
 }
 
 type NetworkConfigList struct {
+	ID           string
 	Name         string
 	CNIVersion   string
 	DisableCheck bool
diff --git a/vendor/github.com/containernetworking/cni/libcni/conf.go b/vendor/github.com/containernetworking/cni/libcni/conf.go
index d8920cf8c..8994bafbf 100644
--- a/vendor/github.com/containernetworking/cni/libcni/conf.go
+++ b/vendor/github.com/containernetworking/cni/libcni/conf.go
@@ -83,6 +83,15 @@ func ConfListFromBytes(bytes []byte) (*NetworkConfigList, error) {
 		}
 	}
 
+	var id string
+	rawId, ok := rawList["id"]
+	if ok {
+		id, ok = rawId.(string)
+		if !ok {
+			return nil, fmt.Errorf("error parsing configuration list: invalid id type %T", rawId)
+		}
+	}
+
 	disableCheck := false
 	if rawDisableCheck, ok := rawList["disableCheck"]; ok {
 		disableCheck, ok = rawDisableCheck.(bool)
@@ -93,6 +102,7 @@ func ConfListFromBytes(bytes []byte) (*NetworkConfigList, error) {
 
 	list := &NetworkConfigList{
 		Name:         name,
+		ID:           id,
 		DisableCheck: disableCheck,
 		CNIVersion:   cniVersion,
 		Bytes:        bytes,

```
if you prefer that I open a PR do not hesitate to ask.

Even if it works I have the feeling that I forgot to modify some parts of the code especially the code related to function `buildOneConfig`.

**Comment by rhatdan** [comment]
Since this involves a rewrite of the network stack, it is not a good first issue.

**Comment by Luap99** [comment]
Fixed by #8549
